src/libertem/common/container.py

```diff
@@ -209,7 +211,17 @@ def _compute_masks(self):
                     default_sparse = False
                 mask_slices.append(m)
 
-        if self._use_sparse is None:
+        if (
+            self._use_sparse is None
+            and default_sparse == 'sparse.pydata'
+            and self.backend in CUPY_BACKENDS
+        ):
+            # The case for shifts, must use sparse.pydata but cannot
+            # run this as sparse on CuPy, so densify to allow calculation
+            # If we reach this, all masks were sparse as default_sparse
+            # remained as 'sparse.pydata' and not False
+            self._use_sparse = False
```

Comment:
> Maybe could have a test that covers this line, just inspecting that `_use_sparse` is disabled if the condition is hit? (or even adding a `using_sparse` property or similar as explicit API for `MaskContainer`)

I think this is actually covered but not in CI - it only applies on a CuPy worker which I don't think we test regularly. Do you want me to test it via some kind of mock ?

Reply:
> I think this is actually covered but not in CI - it only applies on a CuPy worker which I don't think we test regularly. Do you want me to test it via some kind of mock ?

The idea would be a unit test for the `MaskContainer` - don't think that would need to mock things. Maybe it would be good to move the mutation of `self._use_sparse` to `__init__` actually, so that instantiating would be sufficient to check this (and we don't have a weird side-effect in `_compute_masks` anymore...)

Reply:
> (and we don't have a weird side-effect in _compute_masks anymore...)

Agreed, I think there is a lot of scope for refactoring the order of things in `MaskContainer`. That can be part of this PR no problem.

Reply:
![image](https://github.com/LiberTEM/LiberTEM/assets/78845903/eb53ea78-18aa-4b04-8a30-e6a65650a2d7)

While testing this on sparse.pydata backends... Something about the matmul `@` + slice [0] operation causes the sparse matrix to lose all its nonzeros 😱    Will look more into it soon...

Reply:
A workaround for this code was just to use `.reshape()` to drop the singleton first dimension. I spent a bit of time tunneling into the sparse.pydata.matmul code but couldn't figure it out in a reasonable time!

Otherwise I've refactored the internals of `MaskContainer` a bit, it's now a bit more explicit where and how `_use_sparse` is resolved. At a minimum `compute_mask` doesn't do any mutation of attributed, but we still are required to call `compute_mask` to *actually* determine the value of `use_sparse`, as it hinges on the input masks being all sparse, or not.

I've also adapted the array backend support to restrict to compatible backends, we have to exclude the following:

```python
            backends = tuple(
                b for b in self.BACKEND_ALL
                if b not in (
                    self.BACKEND_SCIPY_COO,  # cannot be sliced
                    self.BACKEND_CUPY_SCIPY_COO,  # cannot be sliced
                    self.BACKEND_CUPY_SCIPY_CSR,  # internally cast to BACKEND_CUPY_SCIPY_COO
                    self.BACKEND_CUPY_SCIPY_CSC,  # internally cast to BACKEND_CUPY_SCIPY_COO
                )
            )
```

Reply:
> I spent a bit of time tunneling into the `sparse.pydata.matmul` code but couldn't figure it out in a reasonable time!

Maybe something to report upstream?

> Otherwise I've refactored the internals of MaskContainer a bit

:+1: